src/vms/pvm/etna-builder/spend.ts

```diff
@@ -0,0 +1,473 @@
+import {
+  BigIntPr,
+  Id,
+  Input,
+  OutputOwners,
+  TransferInput,
+  TransferOutput,
+  TransferableInput,
+  TransferableOutput,
+} from '../../../serializable';
+import type { Utxo } from '../../../serializable/avax/utxo';
+import { StakeableLockIn, StakeableLockOut } from '../../../serializable/pvm';
+import { isStakeableLockOut, isTransferOut } from '../../../utils';
+import { matchOwners } from '../../../utils/matchOwners';
+import type { SpendOptions } from '../../common';
+import type { Dimensions } from '../../common/fees/dimensions';
+import type { Serializable } from '../../common/types';
+import type { Context } from '../../context';
+import { SpendHelper } from './spendHelper';
+
+/**
+ * @internal
+ *
+ * Separates the provided UTXOs into two lists:
+ * - `locked` contains UTXOs that have a locktime greater than or equal to `minIssuanceTime`.
+ * - `unlocked` contains UTXOs that have a locktime less than `minIssuanceTime`.
+ *
+ * @param utxos {readonly Utxo[]}
+ * @param minIssuanceTime {bigint}
+ *
+ * @returns Object containing two lists of UTXOs.
+ */
+export const splitByLocktime = (
+  utxos: readonly Utxo[],
+  minIssuanceTime: bigint,
+): { readonly locked: readonly Utxo[]; readonly unlocked: readonly Utxo[] } => {
+  const locked: Utxo[] = [];
+  const unlocked: Utxo[] = [];
+
+  for (const utxo of utxos) {
+    let utxoOwnersLocktime: bigint;
+
+    try {
+      utxoOwnersLocktime = utxo.getOutputOwners().locktime.value();
+    } catch (error) {
+      // If we can't get the locktime, we can't spend the UTXO.
+      // TODO: Is this the right thing to do?
+      // This was necessary to get tests working with testUtxos().
+      continue;
+    }
+
+    if (minIssuanceTime < utxoOwnersLocktime) {
+      locked.push(utxo);
+    } else {
+      unlocked.push(utxo);
+    }
+  }
+
+  return { locked, unlocked };
+};
+
+/**
+ * @internal
+ *
+ * Separates the provided UTXOs into two lists:
+ * - `other` contains UTXOs that have an asset ID different from `assetId`.
+ * - `requested` contains UTXOs that have an asset ID equal to `assetId`.
+ *
+ * @param utxos {readonly Utxo[]}
+ * @param assetId {string}
+ *
+ * @returns Object containing two lists of UTXOs.
+ */
+export const splitByAssetId = (
+  utxos: readonly Utxo[],
+  assetId: string,
+): { readonly other: readonly Utxo[]; readonly requested: readonly Utxo[] } => {
+  const other: Utxo[] = [];
+  const requested: Utxo[] = [];
+
+  for (const utxo of utxos) {
+    if (assetId === utxo.assetId.toString()) {
+      requested.push(utxo);
+    } else {
+      other.push(utxo);
+    }
+  }
+
+  return { other, requested };
+};
+
+/**
+ * @internal
+ *
+ * Returns the TransferOutput that was, potentially, wrapped by a stakeable lockout.
+ *
+ * If the output was stakeable and locked, the locktime is returned.
+ * Otherwise, the locktime returned will be 0n.
+ *
+ * If the output is not an error is returned.
+ */
+export const unwrapOutput = (
```

Comment:
How can this throw? What does that mean? What does this even do? If it is to calculate the stakeableLocktime of the utxo -- can we just use the utility I indicate above?

Reply:
Refactored. Removed this function in favor of the getUtxoInfo util linked.